src/legacy/core_plugins/embeddable_api/public/actions/apply_filter_action.ts

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { i18n } from '@kbn/i18n';
+import { Container, ContainerInput } from '../containers';
+import { IEmbeddable } from '../embeddables';
+import { APPLY_FILTER_TRIGGER, triggerRegistry } from '../triggers';
+import { Filter } from '../types';
+import { Action, ActionContext } from './action';
+import { actionRegistry } from '../actions';
+import { IncompatibleActionError } from './incompatible_action_error';
+import { IContainer } from '../containers/i_container';
+import { attachAction } from '../triggers/attach_action';
+
+interface ApplyFilterContainerInput extends ContainerInput {
+  filters: Filter[];
+}
+
+const APPLY_FILTER_ACTION = 'APPLY_FILTER_ACTION';
+
+function containerAcceptsFilterInput(
+  container: IEmbeddable | IContainer | IContainer<ApplyFilterContainerInput>
+): container is Container<any, ApplyFilterContainerInput> {
+  return (container as Container<any, ApplyFilterContainerInput>).getInput().filters !== undefined;
+}
+
+export class ApplyFilterAction extends Action<IEmbeddable, { filters: Filter[] }> {
+  public readonly type = APPLY_FILTER_ACTION;
+
+  constructor() {
+    super(APPLY_FILTER_ACTION);
+  }
+
+  public getDisplayName() {
+    return i18n.translate('embeddableApi.actions.applyFilterActionTitle', {
+      defaultMessage: 'Apply filter to current view',
+    });
+  }
+
+  public async isCompatible(context: ActionContext) {
+    let root: IEmbeddable | IContainer = context.embeddable;
+    while (root.parent) {
+      root = root.parent;
+    }
+
+    return containerAcceptsFilterInput(root);
+  }
+
+  public execute({
+    embeddable,
+    triggerContext,
+  }: ActionContext<IEmbeddable, { filters: Filter[] }>) {
+    if (!triggerContext) {
+      throw new Error('Applying a filter requires a filter as context');
+    }
+    let root: IEmbeddable | IContainer = embeddable;
+    while (root.parent) {
+      root = root.parent;
+    }
+
+    if (!containerAcceptsFilterInput(root)) {
+      throw new IncompatibleActionError();
+    }
+
+    root.updateInput({
```

Comment:
Yea, that's what I want, for this action.

>  This is not always the behavior you're going to want.

agreed, but if you want the filters to only apply on the specific embeddable, we need to write a custom action that just does `embeddable.updateInput( {filters });` but that would only change the embeddable itself (e.g. a per panel override, stored in `explicitInput`). 

This is how per panel time range will work.  it will do `embeddable.updateInput( { timeRange });`.  When we have both actions available (think, if I brush a chart that has a time range override, do I want to apply it on this embeddable, or at the root parent?  We could offer two actions `apply time range on panel`, vs `apply time range on parent`.  I leave it up to the action to decide, not the listeners of the event, because how would they know what to do? whether to apply directly, or on a parent?

Reply:
just to clarify, so apply_filter action will work both with embeddable (for example discover) as well as with container (dashboard) ?